labcontrol/db/process.py

```diff
@@ -3819,3 +3826,407 @@ def _generate_amplicon_prep_information(self):
+                    primersetcp.primer_set_composition_id)
+                LEFT JOIN labman.primer_set_composition primersetcp2 ON (
+                    primercp2.primer_set_composition_id =
+                    primersetcp2.primer_set_composition_id)
+                LEFT JOIN labman.composition primersetcpcp ON (
+                    primersetcp.composition_id = primersetcpcp.composition_id)
+                LEFT JOIN labman.composition primersetcpcp2 ON (
+                    primersetcp2.composition_id =
+                    primersetcpcp2.composition_id)
+                LEFT JOIN labman.well primersetwell ON (
+                    primersetcpcp.container_id = primersetwell.container_id)
+                LEFT JOIN labman.well primersetwell2 ON (
+                    primersetcpcp2.container_id = primersetwell2.container_id)
+                LEFT JOIN labman.plate primersetplate ON (
+                    primersetwell.plate_id = primersetplate.plate_id)
+                LEFT JOIN labman.plate primersetplate2 ON (
+                    primersetwell2.plate_id = primersetplate2.plate_id)
+                FULL JOIN qiita.study_sample USING (sample_id)
+                LEFT JOIN qiita.study as study USING (study_id)
+                WHERE libprepplate.plate_id IN (
+                    SELECT distinct libprepplate2.plate_id
+                    FROM labman.sequencing_process sp
+                    LEFT JOIN labman.sequencing_process_lanes spl USING (
+                        sequencing_process_id)
+                    LEFT JOIN labman.pool_composition_components pcc1 ON (
+                        spl.pool_composition_id =
+                        pcc1.output_pool_composition_id)
+                   LEFT JOIN labman.library_prep_shotgun_composition libprepcp2
+                        ON (
+                        pcc1.input_composition_id = libprepcp2.composition_id)
+                    LEFT JOIN labman.composition libprepcpcp2 ON (
+                        libprepcp2.composition_id =
+                        libprepcpcp2.composition_id)
+                    LEFT JOIN labman.library_prep_shotgun_process libpreppr2
+                        ON (libprepcpcp2.upstream_process_id =
+                        libpreppr2.process_id)
+                    LEFT JOIN labman.well libprepwell2 ON (
+                        libprepcpcp2.container_id = libprepwell2.container_id)
+                    LEFT JOIN labman.plate libprepplate2 ON (
+                        libprepwell2.plate_id = libprepplate2.plate_id)
+                    WHERE sequencing_process_id = %s)
+                """
+
+        with sql_connection.TRN as TRN:
+            TRN.add(sql, [self.id])
+
+            results = [dict(r) for r in TRN.execute_fetchindex()]
+
+            for d in results:
+                # assert that study_id and orig_name are only None when both
+                # are None; making this a control entry and not a sample entry
```

Comment:
So, the problem is that this isn't necessarily true--at least not currently.  We definitely DO have real, experimental samples coming through and getting plated that happen NOT to be in qiita (which is all that these two pieces of being None indicates).  What is this field being used for?  Should we name it more correct, like "is_not_in_qiita" or something?  I wouldn't want to use it for anything that required that its contents really, truly be controls ...

Reply:
It looks like the reason this line works, at least for the test data, is that the raw results from the SQL query can be organized into two groups. The first group is 288 in number, looks like sample data, and has no missing values.

The second group number 92 and uniformly have NULL values for the following five columns:
['orig_name', 'study_id', 'sample_id', 'project_name', 'experiment_design_description']

Hence, rows where both study_id and 'orig_name' are NULL matches all of the second set and none of the first.

It appears I arrived at this by negating the condition we use to test for a 'sample row' in Amplicon prep info file generation:
`if result['orig_name2'] is not None and study_id is not None:`
Note that orig_name2 is the same as orig_name at this point in the code, and both orig_name and study_id have not been modified since coming off the DB. However, I got it wrong; the negation should be if orig_name is None OR study_id is None.

With respect to the test data, it's amusing that both statements will work equally well in this particular case. IMHO, we could have an is_control column if determining which row represents a control is potentially fuzzy. I could also determine what is a control (or at least not a sample) by parsing the content column, and targeting rows with content starting with 'vibrio' or 'blank' or another word in the table.

I'll leave this one alone until @AmandaBirmingham comments, and then I can make the change easily enough.

Reply:
Thanks, @charles-cowart

Reply:
Please 
(a) take the creation of the `is_control` field out.  It is not clearly named and thus has the potential to lead to gnarly bugs down the road when someone inaccurately assumes it is what it says it is (especially since our existing test conditions don't exercise the ways in which it secretly isn't what it says it is).  Since it is currently used in only one place, we have the option to just drop it until we come up with a better naming/handling of this weird "maybe controls but maybe just unloved experimental samples" situation.
(b) replace the current use of `is_control` to decide when to modify `orig_name2` with the check we use for that in `generate_amplicon_prep_info`:

`if result['orig_name2'] is not None and study_id is not None:`

As I recall, it took you and me about a week of back-and-forth to get this condition *right*, and so I don't want to mess with it again (at least not till we do a fuller refactor someday).

Reply:
Too true!  😆 

Fair enough, although it seems like we could really benefit from easily identifying states. It is something I hope we can get back to at some point.